hi-ml-cpath/src/health_cpath/models/encoders.py

```diff
@@ -86,32 +109,264 @@ def _get_encoder(self) -> Tuple[torch.nn.Module, int]:
         return setup_feature_extractor(pretrained_model, self.input_dim)
 
 
-class Resnet18(ImageNetEncoder):
-    def __init__(self, tile_size: int, n_channels: int = 3) -> None:
-        super().__init__(resnet18, tile_size, n_channels, apply_imagenet_preprocessing=True)
+class ResNetCheckpointingMixin:
+    """Mixin class for checkpointing activations in ResNet-based encoders."""
+
+    def __init__(
+        self,
+        feature_extractor_fn: ResNet,
+        bn_momentum: Optional[float] = None,
+        checkpoint_segments_size: int = 2,
+        checkpoint_activations: bool = False,
+    ) -> None:
+        """
+        :param feature_extractor_fn: A ResNet model.
+        :param bn_momentum: An optional momentum value to use for bn layers statistics updates when
+            checkpoint_activations is True. If None, sqrt of the default momentum is used.
+        :param checkpoint_segments_size: If checkpointing, the size of checkpointed segments in sequential layers
+            (default=2).
+        :param checkpoint_activations: Whether to checkpoint activations during forward pass. This can be used to trade
+            compute for memory. Note that this is only supported for some encoders.
+        """
+        assert isinstance(feature_extractor_fn, ResNet), "Expected ResNet model for feature_extractor_fn argument."
+        self.feature_extractor_fn = feature_extractor_fn
+        self.checkpoint_segments_size = checkpoint_segments_size
+        self.bn_momentum = bn_momentum
+        if checkpoint_activations:
+            self._set_batch_norm_momentum()
+
+    def _set_batch_norm_momentum(self) -> None:
+        """Set the momentum of batch norm layers in the ResNet model to avoid running statistics from going out of
+        sync due to activations checkpointing. The forward pass is applied twice which results in double updates of
+        these statistics. We can workaround that by using sqrt of default momentum.
+        """
+        if self.bn_momentum is not None:
+            _momentum = self.bn_momentum
+        else:
+            _momentum = math.sqrt(self.feature_extractor_fn.bn1.momentum)
+            self.bn_momentum = _momentum
+
+        # Set momentum for the first batch norm layer
+        self.feature_extractor_fn.bn1.momentum = _momentum
+
+        def _set_bn_momentum(layer_block: nn.Sequential) -> None:
+            for sub_layer in layer_block:
+                for key, layer in sub_layer._modules.items():
+                    if "bn" in key:
+                        assert isinstance(layer, nn.BatchNorm2d), "Expected BatchNorm2d layer."
+                        layer.momentum = _momentum
```

Comment:
Maybe overcautious, but would be safer to loop over the layers and modify only those where `isinstance(layer, nn.Batchnorm2d)`? ie, not relying on the naming scheme?

Reply:
good point, done!